include/rapidjson/document.h

```diff
@@ -1298,7 +1429,10 @@ class GenericValue {
 
     //! Set this value as an empty array.
     /*! \post IsArray == true */
-    GenericValue& SetArray() {  this->~GenericValue(); new (this) GenericValue(kArrayType); return *this; }
+    GenericValue& SetArray() { this->~GenericValue(); new (this) GenericValue(kArrayType); return *this; }
+
+    //! Set this value with an array.
+    GenericValue& SetArray(Array& a) { return *this = *a.ptr_; }

```

Comment:
After this call, `a.ptr_` is null, not an (empty) array. We should add an `a.ptr_->SetArray();` here (set for SetObject(Object);` to avoid breaking the invariant.

Secondly, we should check for a.ptr_ not being NULL in an assert, which could happen due to a default constructed `Array|Object` instance.